File: mobile/src/main/java/br/com/helpdev/velocimetroalerta/MainActivity.java
package br.com.helpdev.velocimetroalerta;

import android.Manifest;
import android.content.ComponentName;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.ActivityCompat;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.widget.Toast;

import br.com.helpdev.velocimetroalerta.dialogs.ConfirmDialogFrag;
import br.com.helpdev.velocimetroalerta.gps.ServiceVelocimetro;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener, ServiceConnection {

    private static final int REQUEST_CONFIG_AUDIO = 2;
    private static final int REQUEST_MY_ACTIVIES = 29;
    private static final String SP_KEEP_ALIVE = "keep_alive";
    private static final String LOG = "MainActivity";

    public interface CallbackNotify {
        void onServiceConnected(ServiceVelocimetro serviceVelocimetro);

        void onBeforeDisconnect(ServiceVelocimetro serviceVelocimetro);

        void onCloseProgram();
    }

    private CallbackNotify callbackNotify;
    private ServiceVelocimetro serviceVelocimetro;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        if (savedInstanceState == null) {
            loadPermissions();
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void loadPermissions() {
        if (isPermissionsGranted()) {
            Intent service = new Intent(this, ServiceVelocimetro.class);
            startService(service);
            bindService(service, this, BIND_AUTO_CREATE);
        } else {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.ACCESS_FINE_LOCATION}, 12);
        }
    }

    public boolean isPermissionsGranted() {
        return (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED)
                &&
                (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        switch (id) {
            case R.id.nav_star:
                openAppInGooglePlay();
                break;
            case R.id.nav_atividades:
                startActivityForResult(new Intent(this, MyActivities.class), REQUEST_MY_ACTIVIES);
                break;
            case R.id.nav_config:
                startActivityForResult(new Intent(this, ConfigActivity.class), REQUEST_CONFIG_AUDIO);
                break;
            case R.id.nav_share:
                shareApp();
                break;
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return false;
    }

    private void openAppInGooglePlay() {
        final String appPackageName = getPackageName();
        try {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
        } catch (android.content.ActivityNotFoundException e) { // if there is no Google Play on device
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
        }
    }

    private void shareApp() {
        Intent sendIntent = new Intent();
        sendIntent.setAction(Intent.ACTION_SEND);
        sendIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.app_name) + " https://play.google.com/store/apps/details?id=" + getPackageName());
        sendIntent.setType("text/plain");
        startActivity(sendIntent);
    }

    private SharedPreferences getSharePref() {
        return getSharedPreferences("sp_main_activity", MODE_PRIVATE);
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
            return;
        }

        if (serviceVelocimetro != null && serviceVelocimetro.isRunning()) {
            if (getSupportFragmentManager().findFragmentByTag("DIALOG") == null) {
                ConfirmDialogFrag.getInstance(getString(R.string.confirme_sair), true, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        if (which == DialogInterface.BUTTON_POSITIVE) {
                            stopService();
                            MainActivity.this.finish();
                        }
                    }
                }).show(getSupportFragmentManager(), "DIALOG");
            }
            return;
        } else {
            stopService();
        }

        super.onBackPressed();
    }

    private void stopService() {
        if (callbackNotify != null) callbackNotify.onCloseProgram();
        stopService(new Intent(this, ServiceVelocimetro.class));
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        for (int i : grantResults) {
            if (i != PackageManager.PERMISSION_GRANTED) {
                loadPermissions();
                break;
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        loadIconKeepAlive(menu.getItem(0));
        return true;
    }

    private void loadIconKeepAlive(MenuItem item) {
        if (getSharePref().getBoolean(SP_KEEP_ALIVE, false)) {
            item.setIcon(R.drawable.ic_settings_brightness_black_48dp);
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        } else {
            item.setIcon(R.drawable.ic_settings_brightness_black_48dp_off);
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            startActivityForResult(new Intent(this, ConfigActivity.class), REQUEST_CONFIG_AUDIO);
        } else if (id == R.id.action_keep_alive) {
            String msg;
            if (getSharePref().getBoolean(SP_KEEP_ALIVE, false)) {
                getSharePref().edit().putBoolean(SP_KEEP_ALIVE, false).apply();
                msg = getString(R.string.bloqueio_ativado);
            } else {
                getSharePref().edit().putBoolean(SP_KEEP_ALIVE, true).apply();
                msg = getString(R.string.bloqueio_desativado);
            }
            Toast.makeText(this, msg, Toast.LENGTH_LONG).show();
            loadIconKeepAlive(item);
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (isPermissionsGranted()) {
            bindService(new Intent(this, ServiceVelocimetro.class), this, BIND_AUTO_CREATE);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (serviceVelocimetro != null) {
            try {
                if (callbackNotify != null) callbackNotify.onBeforeDisconnect(serviceVelocimetro);
            } catch (Throwable t) {
                Log.e(LOG, "onBeeforeDisconnect");
            }
            try {
                unbindService(this);
            } catch (Throwable t) {
                Log.e(LOG, "unbindService");
            }
        }
    }

    public ServiceVelocimetro getServiceVelocimetro() {
        return serviceVelocimetro;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CONFIG_AUDIO) {
            configureGpsVelocimetro();
        }
    }

    public void configureGpsVelocimetro() {
        if (serviceVelocimetro != null) {
            serviceVelocimetro.configure();
        }
    }


    public void setCallbackNotify(CallbackNotify callbackNotify) {
        this.callbackNotify = callbackNotify;
    }

    @Override
    public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
        if (iBinder instanceof ServiceVelocimetro.MyBinder) {
            serviceVelocimetro = ((ServiceVelocimetro.MyBinder) iBinder).getServiceVelocimetro();
            configureGpsVelocimetro();
            if (callbackNotify != null) callbackNotify.onServiceConnected(serviceVelocimetro);
        }
    }

    @Override
    public void onServiceDisconnected(ComponentName componentName) {
        serviceVelocimetro = null;
    }
}

File: mobile/src/main/java/br/com/helpdev/velocimetroalerta/MainActivity.kt
package br.com.helpdev.velocimetroalerta

import android.Manifest
import android.content.ComponentName
import android.content.DialogInterface
import android.content.Intent
import android.content.ServiceConnection
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.support.design.widget.NavigationView
import android.support.v4.app.ActivityCompat
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.WindowManager
import android.widget.Toast

import br.com.helpdev.velocimetroalerta.dialogs.ConfirmDialogFrag
import br.com.helpdev.velocimetroalerta.gps.SpeedometerService

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, ServiceConnection {

    private var callbackNotify: CallbackNotify? = null
    var speedometerService: SpeedometerService? = null
        private set

    private val isPermissionsGranted: Boolean
        get() = ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED

    private val sharePref: SharedPreferences
        get() = getSharedPreferences("sp_main_activity", MODE_PRIVATE)

    interface CallbackNotify {
        fun onServiceConnected(speedometerService: SpeedometerService?)

        fun onBeforeDisconnect(speedometerService: SpeedometerService)

        fun onCloseProgram()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        if (savedInstanceState == null) {
            loadPermissions()
        }

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
    }

    private fun loadPermissions() {
        if (isPermissionsGranted) {
            val service = Intent(this, SpeedometerService::class.java)
            startService(service)
            bindService(service, this, BIND_AUTO_CREATE)
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.ACCESS_FINE_LOCATION), 12)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId

        when (id) {
            R.id.nav_star -> openAppInGooglePlay()
            R.id.nav_atividades -> startActivityForResult(Intent(this, MyActivities::class.java), REQUEST_MY_ACTIVITIES)
            R.id.nav_config -> startActivityForResult(Intent(this, ConfigActivity::class.java), REQUEST_CONFIG_AUDIO)
            R.id.nav_share -> shareApp()
        }

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        drawer.closeDrawer(GravityCompat.START)
        return false
    }

    private fun openAppInGooglePlay() {
        val appPackageName = packageName
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$appPackageName")))
        } catch (e: android.content.ActivityNotFoundException) { // if there is no Google Play on device
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=$appPackageName")))
        }

    }

    private fun shareApp() {
        val sendIntent = Intent()
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.app_name) + " https://play.google.com/store/apps/details?id=" + packageName)
        sendIntent.type = "text/plain"
        startActivity(sendIntent)
    }

    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
            return
        }

        if (speedometerService != null && speedometerService!!.isRunning) {
            if (supportFragmentManager.findFragmentByTag("DIALOG") == null) {
                ConfirmDialogFrag.getInstance(getString(R.string.confirme_sair), true, DialogInterface.OnClickListener { dialog, which ->
                    if (which == DialogInterface.BUTTON_POSITIVE) {
                        stopService()
                        this@MainActivity.finish()
                    }
                }).show(supportFragmentManager, "DIALOG")
            }
            return
        } else {
            stopService()
        }

        super.onBackPressed()
    }

    private fun stopService() {
        if (callbackNotify != null) callbackNotify!!.onCloseProgram()
        stopService(Intent(this, SpeedometerService::class.java))
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        for (i in grantResults) {
            if (i != PackageManager.PERMISSION_GRANTED) {
                loadPermissions()
                break
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        loadIconKeepAlive(menu.getItem(0))
        return true
    }

    private fun loadIconKeepAlive(item: MenuItem) {
        if (sharePref.getBoolean(SP_KEEP_ALIVE, false)) {
            item.setIcon(R.drawable.ic_settings_brightness_black_48dp)
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            item.setIcon(R.drawable.ic_settings_brightness_black_48dp_off)
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_settings) {
            if (speedometerService == null || !speedometerService!!.isRunning || speedometerService!!.isPause) {
                startActivityForResult(Intent(this, ConfigActivity::class.java), REQUEST_CONFIG_AUDIO)
            }else{
                val dialog = AlertDialog.Builder(this)
                dialog.setTitle(R.string.app_name)
                dialog.setMessage(R.string.info_config_in_running)
                dialog.setPositiveButton(R.string.bt_ok, null)
                dialog.create().show()
            }
        } else if (id == R.id.action_keep_alive) {
            val msg = if (sharePref.getBoolean(SP_KEEP_ALIVE, false)) {
                sharePref.edit().putBoolean(SP_KEEP_ALIVE, false).apply()
                getString(R.string.bloqueio_ativado)
            } else {
                sharePref.edit().putBoolean(SP_KEEP_ALIVE, true).apply()
                getString(R.string.bloqueio_desativado)
            }
            Toast.makeText(this, msg, Toast.LENGTH_LONG).show()
            loadIconKeepAlive(item)
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        if (isPermissionsGranted) {
            bindService(Intent(this, SpeedometerService::class.java), this, BIND_AUTO_CREATE)
        }
    }

    override fun onPause() {
        super.onPause()
        if (speedometerService != null) {
            try {
                if (callbackNotify != null) callbackNotify!!.onBeforeDisconnect(speedometerService!!)
            } catch (t: Throwable) {
                Log.e(LOG, "onBeeforeDisconnect")
            }

            try {
                unbindService(this)
            } catch (t: Throwable) {
                Log.e(LOG, "unbindService")
            }

        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CONFIG_AUDIO) {
            reloadPreferences()
        }
    }

    private fun reloadPreferences() {
        if (speedometerService != null) {
            speedometerService!!.reloadPreferences()
        }
    }


    fun setCallbackNotify(callbackNotify: CallbackNotify) {
        this.callbackNotify = callbackNotify
    }

    override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
        if (iBinder is SpeedometerService.MyBinder) {
            speedometerService = iBinder.speedometerService
            reloadPreferences()
            if (callbackNotify != null) callbackNotify!!.onServiceConnected(speedometerService)
        }
    }

    override fun onServiceDisconnected(componentName: ComponentName) {
        speedometerService = null
    }

    companion object {
        private const val REQUEST_CONFIG_AUDIO = 2
        private const val REQUEST_MY_ACTIVITIES = 29
        private const val SP_KEEP_ALIVE = "keep_alive"
        private const val LOG = "MainActivity"
    }
}
